hbase/src/main/java/org/apache/zeppelin/hbase/HbaseInterpreter.java

```diff
@@ -143,30 +156,26 @@ public Scheduler getScheduler() {
         HbaseInterpreter.class.getName() + this.hashCode());
   }
 
-  @Override
-  public List<InterpreterCompletion> completion(String buf, int cursor,
-      InterpreterContext interpreterContext) {
-    return null;
+  private void stopProcess(String paragraphId) {
+    if (runningProcesses.containsKey(paragraphId)) {
+      final Executor executor = runningProcesses.get(paragraphId);
+      final ExecuteWatchdog watchdog = executor.getWatchdog();
+      watchdog.destroyProcess();
+      runningProcesses.remove(paragraphId);
+    }
   }
 
-  private static String getSystemDefault(
-      String envName,
-      String propertyName,
-      String defaultValue) {
-
-    if (envName != null && !envName.isEmpty()) {
-      String envValue = System.getenv().get(envName);
-      if (envValue != null) {
-        return envValue;
-      }
-    }
+  private File createTempFile(String paragraphId) throws IOException {
+    File temp = Files.createTempFile(
+            Paths.get(System.getProperty("java.io.tmpdir"), "zeppelin-hbase-scripts"), paragraphId, ".txt").toFile();
+    tempFiles.put(paragraphId, temp);
+    return temp;
+  }
 
-    if (propertyName != null && !propertyName.isEmpty()) {
-      String propValue = System.getProperty(propertyName);
-      if (propValue != null) {
-        return propValue;
-      }
+  private void deleteTempFile(String paragraphId) {
+    if (tempFiles.containsKey(paragraphId)) {
```

Comment:
More elegant style for the same behavior.
```
    File tmpFile = tempFiles.remove(paragraphId);
    if (tmpFile != null) {
      FileUtils.deleteQuietly(tmpFile);
    }
```

With this code style, the map is only searched once and the program execution is faster.
I know that this map is usually very small and does not play a role on today's hardware. But maybe you have other projects where the maps are very large and so you can optimize your application accordingly.

Reply:
@Reamer ,
Great idea. As the return value of the map's remove method indicates the entry to be removed, we should use it instead, avoiding iterating the map again.

Reply:
That is correct. The adaptation of the algorithm in the `stopProcess` method is still missing.